Idle timeout occured sometimes on HTTP/2 client with `InputStreamResponseListener`

**Jetty version(s)**
11.0.8

**Java version/vendor**
openjdk version "11.0.16" 2022-07-19
OpenJDK Runtime Environment (build 11.0.16+8-post-Ubuntu-0ubuntu118.04)
OpenJDK 64-Bit Server VM (build 11.0.16+8-post-Ubuntu-0ubuntu118.04, mixed mode, sharing)

**OS type/version**
Linux 5.4.0-125-generic 141~18.04.1-Ubuntu  x86_64 x86_64 x86_64 GNU/Linux

**Description**
We have switched in production the response listener of the HTTP/2 client, before we used the `FutureResponseListener` and now the `InputStreamResponseListener`, since this change we are observing regularly failures with timeouts and it happens randomly.

This is very similar in symptoms to the issue https://github.com/eclipse/jetty.project/issues/7192

**How to reproduce?**
The problem is difficult to reproduce due to its randomness. As described in the [comment](https://github.com/eclipse/jetty.project/issues/7192#issuecomment-991068558) of the similar issue (https://github.com/eclipse/jetty.project/issues/7192), when we increase the log level to `DEBUG`, the problem is even less reproducible.

I still managed to reproduce the problem thanks to a test in the jetty project that triggers the timeout almost every time.

<details><summary>Test case - Code</summary>
<p>

```java
public class HttpClientTransportOverHTTP2Test extends AbstractTest
{

    private static final Logger LOG = LoggerFactory.getLogger(HttpClientTransportOverHTTP2Test.class);

    @Test
    public void testInputStreamResponseListener() throws Exception
    {
        var bytes = 100_000;
        start(new ServerSessionListener.Adapter()
        {
            @Override
            public Stream.Listener onNewStream(Stream stream, HeadersFrame frame)
            {
                int streamId = stream.getId();
                MetaData.Response response = new MetaData.Response(HttpVersion.HTTP_2, HttpStatus.OK_200, HttpFields.EMPTY);
                HeadersFrame responseFrame = new HeadersFrame(streamId, response, null, false);
                Callback.Completable callback = new Callback.Completable();
                stream.headers(responseFrame, callback);
                callback.thenRun(() -> stream.data(new DataFrame(streamId, ByteBuffer.wrap(new byte[bytes]), true), Callback.NOOP));
                return null;
            }
        });

        var nbRequests = 100_000;
        var progress = new AtomicInteger(0);
        IntStream.range(0, nbRequests).forEach(i -> {
            try {
                if (progress.incrementAndGet() % 1000 == 0) {
                    LOG.info("progress {}/{}", progress.get(), nbRequests);
                }
                InputStreamResponseListener listener = new InputStreamResponseListener();
                client.newRequest("localhost", connector.getLocalPort()).headers(httpFields -> httpFields.put("X-Request-Id", Integer.toString(i))).send(listener);
                Response response = listener.get(60, TimeUnit.SECONDS);
                assertEquals(HttpStatus.OK_200, response.getStatus());
                assertEquals(bytes, listener.getInputStream().readAllBytes().length);
            } catch (Exception e) {
                LOG.error("request {} failed: {}", i, e.getMessage());
                fail(e);
            }
        });
    }

}
```

</p>
</details>

<details>
<summary>Test case - Failure Stacktrace</summary>
<p>

```java
org.opentest4j.AssertionFailedError
	at org.junit.jupiter.api.AssertionUtils.fail(AssertionUtils.java:47)
	at org.junit.jupiter.api.Assertions.fail(Assertions.java:157)
	at org.eclipse.jetty.http2.http.client.transport/org.eclipse.jetty.http2.client.http.HttpClientTransportOverHTTP2Test.lambda$testInputStreamResponseListener$1(HttpClientTransportOverHTTP2Test.java:77)
	at java.base/java.util.stream.Streams$RangeIntSpliterator.forEachRemaining(Streams.java:104)
	at java.base/java.util.stream.IntPipeline$Head.forEach(IntPipeline.java:593)
	at org.eclipse.jetty.http2.http.client.transport/org.eclipse.jetty.http2.client.http.HttpClientTransportOverHTTP2Test.testInputStreamResponseListener(HttpClientTransportOverHTTP2Test.java:65)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:64)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:725)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:149)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:140)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:84)
	at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$7(TestMethodTestDescriptor.java:214)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:210)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:135)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:66)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:151)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:107)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86)
	at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)
	at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:53)
	at com.intellij.junit5.JUnit5IdeaTestRunner.startRunnerWithArgs(JUnit5IdeaTestRunner.java:71)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater$1.execute(IdeaTestRunner.java:38)
	at com.intellij.rt.execution.junit.TestsRepeater.repeat(TestsRepeater.java:11)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:35)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:235)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:54)
Caused by: java.io.IOException: java.util.concurrent.TimeoutException: Idle timeout expired: 5000 ms
	at org.eclipse.jetty.client/org.eclipse.jetty.client.util.InputStreamResponseListener$Input.toIOException(InputStreamResponseListener.java:342)
	at org.eclipse.jetty.client/org.eclipse.jetty.client.util.InputStreamResponseListener$Input.read(InputStreamResponseListener.java:310)
	at java.base/java.io.InputStream.readNBytes(InputStream.java:409)
	at java.base/java.io.InputStream.readAllBytes(InputStream.java:346)
	at org.eclipse.jetty.http2.http.client.transport/org.eclipse.jetty.http2.client.http.HttpClientTransportOverHTTP2Test.lambda$testInputStreamResponseListener$1(HttpClientTransportOverHTTP2Test.java:74)
	... 72 more
Caused by: java.util.concurrent.TimeoutException: Idle timeout expired: 5000 ms
	at org.eclipse.jetty.http2.http.client.transport/org.eclipse.jetty.http2.client.http.HTTPSessionListenerPromise.onIdleTimeout(HTTPSessionListenerPromise.java:103)
	at org.eclipse.jetty.http2.common/org.eclipse.jetty.http2.HTTP2Session.notifyIdleTimeout(HTTP2Session.java:1165)
	at org.eclipse.jetty.http2.common/org.eclipse.jetty.http2.HTTP2Session$StreamsState.onIdleTimeout(HTTP2Session.java:1916)
	at org.eclipse.jetty.http2.common/org.eclipse.jetty.http2.HTTP2Session.onIdleTimeout(HTTP2Session.java:1006)
	at org.eclipse.jetty.http2.common/org.eclipse.jetty.http2.HTTP2Connection.onIdleExpired(HTTP2Connection.java:180)
	at org.eclipse.jetty.io/org.eclipse.jetty.io.AbstractEndPoint.onIdleExpired(AbstractEndPoint.java:407)
	at org.eclipse.jetty.io/org.eclipse.jetty.io.IdleTimeout.checkIdleTimeout(IdleTimeout.java:166)
	at org.eclipse.jetty.io/org.eclipse.jetty.io.IdleTimeout.idleCheck(IdleTimeout.java:108)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at java.base/java.lang.Thread.run(Thread.java:832)
```

</p>
</details>

<details>
<summary>Test case - Output</summary>
<p>

#### Full output 
- [jetty_test_case.txt.tar.gz](https://github.com/eclipse/jetty.project/files/9525659/jetty_test_case.txt.tar.gz)

#### Interesting snippets

The failure happened after that the `ContentNotifier` will be marked as `inactive` and `stalled` in this code path :
```java
            // Process only if there is demand.
            try (AutoLock l = lock.lock())
            {
                if (!resume && demand() <= 0)
                {
                    if (LOG.isDebugEnabled())
                        LOG.debug("Stalling processing, content available but no demand");
                    active = false;
                    stalled = true;
                    return;
                }
            }
```
After that the content notifier does not resume the processing and therefore does not notify the `InputStreamResponseListener`, hence the final timeout.
 
```
2022-09-08 11:33:03.461:INFO :oejhch.HttpReceiverOverHTTP2:client-27: [null][ContentNotifier@405c852b] Resuming(false) processing(true) of content [stalled=false]
2022-09-08 11:33:03.461:INFO :oejhch.HttpReceiverOverHTTP2:client-27: [ContentNotifier@405c852b] Stalling processing, content available but no demand
2022-09-08 11:33:03.461:INFO :oejc.HttpReceiver:client-29: [HttpReceiverOverHTTP2@52c3b6de(rsp=TRANSIENT,failure=null)] Response demand=1/1, resume=false [stalled=false]
2022-09-08 11:33:03.461:INFO :oejhch.HttpReceiverOverHTTP2:client-29: [ContentNotifier@405c852b] Queueing content DataInfo@503c7006[DataFrame@2960b028#8493{length:16375,end=false}]
2022-09-08 11:33:03.461:INFO :oejhch.HttpReceiverOverHTTP2:client-29: [4246][ContentNotifier@405c852b] Resuming(false) processing(false) of content [stalled=true]
2022-09-08 11:33:03.461:INFO :oejhch.HttpReceiverOverHTTP2:client-29: [ContentNotifier@405c852b] Queueing content DataInfo@54b2c4c6[DataFrame@79facf75#8493{length:9,end=false}]
2022-09-08 11:33:03.461:INFO :oejhch.HttpReceiverOverHTTP2:client-29: [4246][ContentNotifier@405c852b] Resuming(false) processing(false) of content [stalled=true]
2022-09-08 11:33:03.461:INFO :oejhch.HttpReceiverOverHTTP2:client-29: [ContentNotifier@405c852b] Queueing content DataInfo@4af53c0c[DataFrame@6638348a#8493{length:16366,end=false}]
2022-09-08 11:33:03.461:INFO :oejhch.HttpReceiverOverHTTP2:client-29: [4246][ContentNotifier@405c852b] Resuming(false) processing(false) of content [stalled=true]
2022-09-08 11:33:03.461:INFO :oejhch.HttpReceiverOverHTTP2:client-29: [ContentNotifier@405c852b] Queueing content DataInfo@246e487f[DataFrame@b40aae0#8493{length:18,end=false}]
2022-09-08 11:33:03.461:INFO :oejhch.HttpReceiverOverHTTP2:client-29: [4246][ContentNotifier@405c852b] Resuming(false) processing(false) of content [stalled=true]
2022-09-08 11:33:03.461:INFO :oejhch.HttpReceiverOverHTTP2:client-29: [ContentNotifier@405c852b] Queueing content DataInfo@6836f2d8[DataFrame@4f8d1730#8493{length:16357,end=false}]
2022-09-08 11:33:03.461:INFO :oejhch.HttpReceiverOverHTTP2:client-29: [4246][ContentNotifier@405c852b] Resuming(false) processing(false) of content [stalled=true]
2022-09-08 11:33:03.461:INFO :oejhch.HttpReceiverOverHTTP2:client-29: [ContentNotifier@405c852b] Queueing content DataInfo@19f6930e[DataFrame@176f3d7a#8493{length:27,end=false}]
2022-09-08 11:33:03.461:INFO :oejhch.HttpReceiverOverHTTP2:client-29: [4246][ContentNotifier@405c852b] Resuming(false) processing(false) of content [stalled=true]
2022-09-08 11:33:03.461:INFO :oejhch.HttpReceiverOverHTTP2:client-29: [ContentNotifier@405c852b] Queueing content DataInfo@4890ad7d[DataFrame@6f0307c8#8493{length:16348,end=false}]
2022-09-08 11:33:03.461:INFO :oejhch.HttpReceiverOverHTTP2:client-29: [4246][ContentNotifier@405c852b] Resuming(false) processing(false) of content [stalled=true]
2022-09-08 11:33:03.461:INFO :oejhch.HttpReceiverOverHTTP2:client-29: [ContentNotifier@405c852b] Queueing content DataInfo@424097c5[DataFrame@2caae61a#8493{length:36,end=false}]
2022-09-08 11:33:03.461:INFO :oejhch.HttpReceiverOverHTTP2:client-29: [4246][ContentNotifier@405c852b] Resuming(false) processing(false) of content [stalled=true]
2022-09-08 11:33:03.461:INFO :oejhch.HttpReceiverOverHTTP2:client-29: [ContentNotifier@405c852b] Queueing content DataInfo@619c4db7[DataFrame@16563b0b#8493{length:16339,end=false}]
2022-09-08 11:33:03.461:INFO :oejhch.HttpReceiverOverHTTP2:client-29: [4246][ContentNotifier@405c852b] Resuming(false) processing(false) of content [stalled=true]
2022-09-08 11:33:03.461:INFO :oejhch.HttpReceiverOverHTTP2:client-29: [ContentNotifier@405c852b] Queueing content DataInfo@506e68bb[DataFrame@2b0fce9a#8493{length:45,end=false}]
2022-09-08 11:33:03.461:INFO :oejhch.HttpReceiverOverHTTP2:client-29: [4246][ContentNotifier@405c852b] Resuming(false) processing(false) of content [stalled=true]
2022-09-08 11:33:03.461:INFO :oejhch.HttpReceiverOverHTTP2:client-29: [ContentNotifier@405c852b] Queueing content DataInfo@287ddfb9[DataFrame@fd437de#8493{length:16330,end=false}]
2022-09-08 11:33:03.461:INFO :oejhch.HttpReceiverOverHTTP2:client-29: [4246][ContentNotifier@405c852b] Resuming(false) processing(false) of content [stalled=true]
2022-09-08 11:33:03.461:INFO :oejhch.HttpReceiverOverHTTP2:client-29: [ContentNotifier@405c852b] Queueing content DataInfo@2cdcad2f[DataFrame@7972cd30#8493{length:54,end=false}]
2022-09-08 11:33:03.461:INFO :oejhch.HttpReceiverOverHTTP2:client-29: [4246][ContentNotifier@405c852b] Resuming(false) processing(false) of content [stalled=true]
2022-09-08 11:33:03.461:INFO :oejhch.HttpReceiverOverHTTP2:client-29: [ContentNotifier@405c852b] Queueing content DataInfo@3f3445b0[DataFrame@11f0ef4f#8493{length:1696,end=true}]
2022-09-08 11:33:03.461:INFO :oejhch.HttpReceiverOverHTTP2:client-29: [4246][ContentNotifier@405c852b] Resuming(false) processing(false) of content [stalled=true]
2022-09-08 11:33:08.469:INFO :oejc.HttpReceiver:HttpClient@13e344d-scheduler-1: cleanup HttpReceiverOverHTTP2@52c3b6de(rsp=FAILURE,failure=java.util.concurrent.TimeoutException: Idle timeout expired: 5000 ms)
2022-09-08 11:33:08.469:ERROR:oejhch.HttpClientTransportOverHTTP2Test:main: request 4246 failed: java.util.concurrent.TimeoutException: Idle timeout expired: 5000 ms
```
After that my understanding of the code and the design of the components is not sufficient to go further in the analysis. I don't know if I'm completely on the wrong track or not, WDYT ?

</p>
</details>

--- a/jetty-http2/http2-common/src/main/java/org/eclipse/jetty/http2/HTTP2Stream.java
+++ b/jetty-http2/http2-common/src/main/java/org/eclipse/jetty/http2/HTTP2Stream.java
@@ -285,6 +285,15 @@ public class HTTP2Stream implements IStream, Callback, Dumpable, CyclicTimeouts.
         return committed;
     }
 
+    @Override
+    public int dataSize()
+    {
+        try (AutoLock l = lock.lock())
+        {
+            return dataQueue == null ? 0 : dataQueue.size();
+        }
+    }
+
     public boolean isOpen()
     {
         return !isClosed();
@@ -921,13 +930,14 @@ public class HTTP2Stream implements IStream, Callback, Dumpable, CyclicTimeouts.
     @Override
     public String toString()
     {
-        return String.format("%s@%x#%d@%x{sendWindow=%s,recvWindow=%s,demand=%d,reset=%b/%b,%s,age=%d,attachment=%s}",
+        return String.format("%s@%x#%d@%x{sendWindow=%s,recvWindow=%s,queue=%d,demand=%d,reset=%b/%b,%s,age=%d,attachment=%s}",
             getClass().getSimpleName(),
             hashCode(),
             getId(),
             session.hashCode(),
             sendWindow,
             recvWindow,
+            dataSize(),
             demand(),
             localReset,
             remoteReset,

--- a/jetty-http2/http2-common/src/main/java/org/eclipse/jetty/http2/IStream.java
+++ b/jetty-http2/http2-common/src/main/java/org/eclipse/jetty/http2/IStream.java
@@ -143,6 +143,11 @@ public interface IStream extends Stream, Attachable, Closeable
     boolean isCommitted();
 
     /**
+     * @return the size of the DATA frame queue
+     */
+    int dataSize();
+
+    /**
      * <p>An ordered list of frames belonging to the same stream.</p>
      */
     public static class FrameList

--- a/jetty-http2/http2-http-client-transport/src/main/java/org/eclipse/jetty/http2/client/http/HttpChannelOverHTTP2.java
+++ b/jetty-http2/http2-http-client-transport/src/main/java/org/eclipse/jetty/http2/client/http/HttpChannelOverHTTP2.java
@@ -186,7 +186,14 @@ public class HttpChannelOverHTTP2 extends HttpChannel
         }
 
         @Override
-        public void onData(Stream stream, DataFrame frame, Callback callback)
+        public void onBeforeData(Stream stream)
+        {
+            // Don't demand here, as the initial demand is controlled by
+            // the application via DemandedContentListener.onBeforeContent().
+        }
+
+        @Override
+        public void onDataDemanded(Stream stream, DataFrame frame, Callback callback)
         {
             HTTP2Channel.Client channel = (HTTP2Channel.Client)((IStream)stream).getAttachment();
             channel.onData(frame, callback);

--- a/jetty-http2/http2-http-client-transport/src/main/java/org/eclipse/jetty/http2/client/http/HttpReceiverOverHTTP2.java
+++ b/jetty-http2/http2-http-client-transport/src/main/java/org/eclipse/jetty/http2/client/http/HttpReceiverOverHTTP2.java
@@ -14,10 +14,7 @@
 package org.eclipse.jetty.http2.client.http;
 
 import java.io.IOException;
-import java.nio.ByteBuffer;
-import java.util.ArrayDeque;
 import java.util.List;
-import java.util.Queue;
 import java.util.function.BiFunction;
 
 import org.eclipse.jetty.client.HttpChannel;
@@ -42,9 +39,7 @@ import org.eclipse.jetty.http2.frames.HeadersFrame;
 import org.eclipse.jetty.http2.frames.PushPromiseFrame;
 import org.eclipse.jetty.http2.frames.ResetFrame;
 import org.eclipse.jetty.io.EndPoint;
-import org.eclipse.jetty.util.BufferUtil;
 import org.eclipse.jetty.util.Callback;
-import org.eclipse.jetty.util.thread.AutoLock;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -68,7 +63,11 @@ public class HttpReceiverOverHTTP2 extends HttpReceiver implements HTTP2Channel.
     @Override
     protected void receive()
     {
-        contentNotifier.process(true);
+        HttpExchange exchange = getHttpExchange();
+        if (exchange == null)
+            return;
+
+        contentNotifier.receive(getHttpChannel().getStream(), exchange);
     }
 
     @Override
@@ -117,20 +116,14 @@ public class HttpReceiverOverHTTP2 extends HttpReceiver implements HTTP2Channel.
                         upgrade(upgrader, httpResponse, endPoint);
                 }
 
+                contentNotifier.notifySuccess = frame.isEndStream();
                 if (responseHeaders(exchange))
                 {
                     int status = response.getStatus();
                     if (frame.isEndStream() || HttpStatus.isInterim(status))
                         responseSuccess(exchange);
-                }
-                else
-                {
-                    if (frame.isEndStream())
-                    {
-                        // There is no demand to trigger response success, so add
-                        // a poison pill to trigger it when there will be demand.
-                        notifyContent(exchange, new DataFrame(stream.getId(), BufferUtil.EMPTY_BUFFER, true), Callback.NOOP);
-                    }
+                    else
+                        stream.demand(1);
                 }
             }
         }
@@ -138,10 +131,11 @@ public class HttpReceiverOverHTTP2 extends HttpReceiver implements HTTP2Channel.
         {
             HttpFields trailers = metaData.getFields();
             trailers.forEach(httpResponse::trailer);
-            // Previous DataFrames had endStream=false, so
-            // add a poison pill to trigger response success
-            // after all normal DataFrames have been consumed.
-            notifyContent(exchange, new DataFrame(stream.getId(), BufferUtil.EMPTY_BUFFER, true), Callback.NOOP);
+
+            if (((IStream)stream).dataSize() == 0)
+                responseSuccess(exchange);
+            else
+                contentNotifier.notifySuccess = true;
         }
     }
 
@@ -194,13 +188,9 @@ public class HttpReceiverOverHTTP2 extends HttpReceiver implements HTTP2Channel.
     {
         HttpExchange exchange = getHttpExchange();
         if (exchange == null)
-        {
             callback.failed(new IOException("terminated"));
-        }
         else
-        {
             notifyContent(exchange, frame, callback);
-        }
     }
 
     void onReset(Stream stream, ResetFrame frame)
@@ -230,198 +220,68 @@ public class HttpReceiverOverHTTP2 extends HttpReceiver implements HTTP2Channel.
 
     private void notifyContent(HttpExchange exchange, DataFrame frame, Callback callback)
     {
-        contentNotifier.offer(exchange, frame, callback);
+        if (LOG.isDebugEnabled())
+            LOG.debug("Received content {}", frame);
+        contentNotifier.process(getHttpChannel().getStream(), exchange, frame, callback);
     }
 
-    private class ContentNotifier
+    private static class ContentNotifier
     {
-        private final AutoLock lock = new AutoLock();
-        private final Queue<DataInfo> queue = new ArrayDeque<>();
         private final HttpReceiverOverHTTP2 receiver;
-        private DataInfo dataInfo;
-        private boolean active;
-        private boolean resume;
-        private boolean stalled;
+        private volatile boolean notifySuccess;
 
         private ContentNotifier(HttpReceiverOverHTTP2 receiver)
         {
             this.receiver = receiver;
         }
 
-        private void offer(HttpExchange exchange, DataFrame frame, Callback callback)
+        public void receive(Stream stream, HttpExchange exchange)
         {
-            DataInfo dataInfo = new DataInfo(exchange, frame, callback);
-            if (LOG.isDebugEnabled())
-                LOG.debug("Queueing content {}", dataInfo);
-            enqueue(dataInfo);
-            process(false);
+            if (notifySuccess && ((IStream)stream).dataSize() == 0)
+                receiver.responseSuccess(exchange);
+            else
+                stream.demand(1);
         }
 
-        private void enqueue(DataInfo dataInfo)
+        private void process(Stream stream, HttpExchange exchange, DataFrame dataFrame, Callback callback)
         {
-            try (AutoLock l = lock.lock())
+            if (dataFrame.getData().hasRemaining())
             {
-                queue.offer(dataInfo);
-            }
-        }
-
-        private void process(boolean resume)
-        {
-            // Allow only one thread at a time.
-            boolean busy = active(resume);
-            if (LOG.isDebugEnabled())
-                LOG.debug("Resuming({}) processing({}) of content", resume, !busy);
-            if (busy)
-                return;
-
-            // Process only if there is demand.
-            try (AutoLock l = lock.lock())
-            {
-                if (!resume && demand() <= 0)
+                if (dataFrame.isEndStream())
+                    notifySuccess = true;
+                boolean proceed = receiver.responseContent(exchange, dataFrame.getData(), Callback.from(callback::succeeded, x -> fail(callback, x)));
+                if (proceed)
                 {
-                    if (LOG.isDebugEnabled())
-                        LOG.debug("Stalling processing, content available but no demand");
-                    active = false;
-                    stalled = true;
-                    return;
-                }
-            }
-
-            while (true)
-            {
-                if (dataInfo != null)
-                {
-                    if (dataInfo.frame.isEndStream())
-                    {
-                        receiver.responseSuccess(dataInfo.exchange);
-                        // Return even if active, as reset() will be called later.
-                        return;
-                    }
-                }
-
-                try (AutoLock l = lock.lock())
-                {
-                    dataInfo = queue.poll();
-                    if (LOG.isDebugEnabled())
-                        LOG.debug("Processing content {}", dataInfo);
-                    if (dataInfo == null)
-                    {
-                        active = false;
-                        return;
-                    }
-                }
-
-                ByteBuffer buffer = dataInfo.frame.getData();
-                Callback callback = dataInfo.callback;
-                if (buffer.hasRemaining())
-                {
-                    boolean proceed = receiver.responseContent(dataInfo.exchange, buffer, Callback.from(callback::succeeded, x -> fail(callback, x)));
-                    if (!proceed)
-                    {
-                        // The call to responseContent() said we should
-                        // stall, but another thread may have just resumed.
-                        boolean stall = stall();
-                        if (LOG.isDebugEnabled())
-                            LOG.debug("Stalling({}) processing", stall);
-                        if (stall)
-                            return;
-                    }
+                    if (dataFrame.isEndStream())
+                        receiver.responseSuccess(exchange);
+                    else
+                        stream.demand(1);
                 }
                 else
                 {
-                    callback.succeeded();
+                    if (LOG.isDebugEnabled())
+                        LOG.debug("Stalling processing, no demand after {} on {}", dataFrame, this);
                 }
             }
-        }
-
-        private boolean active(boolean resume)
-        {
-            try (AutoLock l = lock.lock())
+            else
             {
-                if (active)
-                {
-                    // There is a thread in process(),
-                    // but it may be about to exit, so
-                    // remember "resume" to signal the
-                    // processing thread to continue.
-                    if (resume)
-                        this.resume = true;
-                    return true;
-                }
-
-                // If there is no demand (i.e. stalled
-                // and not resuming) then don't process.
-                if (stalled && !resume)
-                    return true;
-
-                // Start processing.
-                active = true;
-                stalled = false;
-                return false;
-            }
-        }
-
-        /**
-         * Called when there is no demand, this method checks whether
-         * the processing should really stop or it should continue.
-         *
-         * @return true to stop processing, false to continue processing
-         */
-        private boolean stall()
-        {
-            try (AutoLock l = lock.lock())
-            {
-                if (resume)
-                {
-                    // There was no demand, but another thread
-                    // just demanded, continue processing.
-                    resume = false;
-                    return false;
-                }
-
-                // There is no demand, stop processing.
-                active = false;
-                stalled = true;
-                return true;
+                callback.succeeded();
+                if (dataFrame.isEndStream())
+                    receiver.responseSuccess(exchange);
+                else
+                    stream.demand(1);
             }
         }
 
         private void reset()
         {
-            dataInfo = null;
-            try (AutoLock l = lock.lock())
-            {
-                queue.clear();
-                active = false;
-                resume = false;
-                stalled = false;
-            }
+            notifySuccess = false;
         }
 
         private void fail(Callback callback, Throwable failure)
         {
             callback.failed(failure);
             receiver.responseFailure(failure);
-        }
-
-        private class DataInfo
-        {
-            private final HttpExchange exchange;
-            private final DataFrame frame;
-            private final Callback callback;
-
-            private DataInfo(HttpExchange exchange, DataFrame frame, Callback callback)
-            {
-                this.exchange = exchange;
-                this.frame = frame;
-                this.callback = callback;
-            }
-
-            @Override
-            public String toString()
-            {
-                return String.format("%s@%x[%s]", getClass().getSimpleName(), hashCode(), frame);
-            }
         }
     }
 }

--- a/jetty-http2/http2-http-client-transport/src/test/java/org/eclipse/jetty/http2/client/http/HttpClientTransportOverHTTP2Test.java
+++ b/jetty-http2/http2-http-client-transport/src/test/java/org/eclipse/jetty/http2/client/http/HttpClientTransportOverHTTP2Test.java
@@ -85,7 +85,6 @@ import static org.junit.jupiter.api.Assertions.assertNotNull;
 import static org.junit.jupiter.api.Assertions.assertSame;
 import static org.junit.jupiter.api.Assertions.assertThrows;
 import static org.junit.jupiter.api.Assertions.assertTrue;
-import static org.junit.jupiter.api.Assertions.fail;
 
 public class HttpClientTransportOverHTTP2Test extends AbstractTest
 {
@@ -634,13 +633,10 @@ public class HttpClientTransportOverHTTP2Test extends AbstractTest
         });
 
         var requestCount = 10_000;
-        var progress = new AtomicInteger(0);
         IntStream.range(0, requestCount).forEach(i ->
         {
             try
             {
-                if (progress.incrementAndGet() % 1000 == 0)
-                    System.err.printf("progress %d/%d%n", progress.get(), requestCount);
                 InputStreamResponseListener listener = new InputStreamResponseListener();
                 client.newRequest("localhost", connector.getLocalPort()).headers(httpFields -> httpFields.put("X-Request-Id", Integer.toString(i))).send(listener);
                 Response response = listener.get(15, TimeUnit.SECONDS);
@@ -649,7 +645,7 @@ public class HttpClientTransportOverHTTP2Test extends AbstractTest
             }
             catch (Exception e)
             {
-                fail(e);
+                throw new RuntimeException(e);
             }
         });
     }

--- a/jetty-http2/http2-http-client-transport/src/test/java/org/eclipse/jetty/http2/client/http/HttpClientTransportOverHTTP2Test.java
+++ b/jetty-http2/http2-http-client-transport/src/test/java/org/eclipse/jetty/http2/client/http/HttpClientTransportOverHTTP2Test.java
@@ -32,6 +32,7 @@ import java.util.concurrent.TimeoutException;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.UnaryOperator;
+import java.util.stream.IntStream;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
@@ -40,6 +41,8 @@ import org.eclipse.jetty.client.HttpDestination;
 import org.eclipse.jetty.client.HttpProxy;
 import org.eclipse.jetty.client.Origin;
 import org.eclipse.jetty.client.api.ContentResponse;
+import org.eclipse.jetty.client.api.Response;
+import org.eclipse.jetty.client.util.InputStreamResponseListener;
 import org.eclipse.jetty.http.HttpFields;
 import org.eclipse.jetty.http.HttpMethod;
 import org.eclipse.jetty.http.HttpStatus;
@@ -82,6 +85,7 @@ import static org.junit.jupiter.api.Assertions.assertNotNull;
 import static org.junit.jupiter.api.Assertions.assertSame;
 import static org.junit.jupiter.api.Assertions.assertThrows;
 import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assertions.fail;
 
 public class HttpClientTransportOverHTTP2Test extends AbstractTest
 {
@@ -608,6 +612,46 @@ public class HttpClientTransportOverHTTP2Test extends AbstractTest
             });
 
         assertTrue(latch.await(5, TimeUnit.SECONDS));
+    }
+
+    @Test
+    public void testInputStreamResponseListener() throws Exception
+    {
+        var bytes = 100_000;
+        start(new ServerSessionListener.Adapter()
+        {
+            @Override
+            public Stream.Listener onNewStream(Stream stream, HeadersFrame frame)
+            {
+                int streamId = stream.getId();
+                MetaData.Response response = new MetaData.Response(HttpVersion.HTTP_2, HttpStatus.OK_200, HttpFields.EMPTY);
+                HeadersFrame responseFrame = new HeadersFrame(streamId, response, null, false);
+                Callback.Completable callback = new Callback.Completable();
+                stream.headers(responseFrame, callback);
+                callback.thenRun(() -> stream.data(new DataFrame(streamId, ByteBuffer.wrap(new byte[bytes]), true), Callback.NOOP));
+                return null;
+            }
+        });
+
+        var requestCount = 10_000;
+        var progress = new AtomicInteger(0);
+        IntStream.range(0, requestCount).forEach(i ->
+        {
+            try
+            {
+                if (progress.incrementAndGet() % 1000 == 0)
+                    System.err.printf("progress %d/%d%n", progress.get(), requestCount);
+                InputStreamResponseListener listener = new InputStreamResponseListener();
+                client.newRequest("localhost", connector.getLocalPort()).headers(httpFields -> httpFields.put("X-Request-Id", Integer.toString(i))).send(listener);
+                Response response = listener.get(15, TimeUnit.SECONDS);
+                assertEquals(HttpStatus.OK_200, response.getStatus());
+                assertEquals(bytes, listener.getInputStream().readAllBytes().length);
+            }
+            catch (Exception e)
+            {
+                fail(e);
+            }
+        });
     }
 
     @Disabled

--- a/jetty-http3/http3-http-client-transport/src/main/java/org/eclipse/jetty/http3/client/http/internal/HttpReceiverOverHTTP3.java
+++ b/jetty-http3/http3-http-client-transport/src/main/java/org/eclipse/jetty/http3/client/http/internal/HttpReceiverOverHTTP3.java
@@ -34,7 +34,7 @@ import org.slf4j.LoggerFactory;
 public class HttpReceiverOverHTTP3 extends HttpReceiver implements Stream.Client.Listener
 {
     private static final Logger LOG = LoggerFactory.getLogger(HttpReceiverOverHTTP3.class);
-    private boolean notifySuccess;
+    private volatile boolean notifySuccess;
 
     protected HttpReceiverOverHTTP3(HttpChannelOverHTTP3 channel)
     {
@@ -86,6 +86,7 @@ public class HttpReceiverOverHTTP3 extends HttpReceiver implements Stream.Client
 
             // TODO: add support for HttpMethod.CONNECT.
 
+            notifySuccess = frame.isLast();
             if (responseHeaders(exchange))
             {
                 int status = response.getStatus();
@@ -98,7 +99,6 @@ public class HttpReceiverOverHTTP3 extends HttpReceiver implements Stream.Client
             {
                 if (LOG.isDebugEnabled())
                     LOG.debug("stalling response processing, no demand after headers on {}", this);
-                notifySuccess = frame.isLast();
             }
         }
     }
@@ -118,12 +118,16 @@ public class HttpReceiverOverHTTP3 extends HttpReceiver implements Stream.Client
                 ByteBuffer byteBuffer = data.getByteBuffer();
                 if (byteBuffer.hasRemaining())
                 {
+                    if (data.isLast())
+                        notifySuccess = true;
+
                     Callback callback = Callback.from(Invocable.InvocationType.NON_BLOCKING, data::complete, x ->
                     {
                         data.complete();
                         if (responseFailure(x))
                             stream.reset(HTTP3ErrorCode.REQUEST_CANCELLED_ERROR.code(), x);
                     });
+
                     boolean proceed = responseContent(exchange, byteBuffer, callback);
                     if (proceed)
                     {
@@ -136,7 +140,6 @@ public class HttpReceiverOverHTTP3 extends HttpReceiver implements Stream.Client
                     {
                         if (LOG.isDebugEnabled())
                             LOG.debug("stalling response processing, no demand after {} on {}", data, this);
-                        notifySuccess = data.isLast();
                     }
                 }
                 else

